File: cracking_the_coding_interview/stack_queue/queue.php
<?php

class Node {
  
  function __consturct($data){
    $this->data = $data;
    $this->next = null;
  }

}
Class Queue {

  public $queue;
  public $rear = null;
  public $front = null;

  function __construct(){
    $this->queue = [];
  }

  public function enqueue($data){

    $node = new Node($data);
    $node->data = $data;
    if($this->rear){
      $this->rear->next = $node;
    }
    $this->rear = $node;

    if($this->front == null){
      $this->front = $node;
    }

    echo $data." is enqued. ";

  }

  public function dequeue(){
    if($this->front){
      $front = $this->front;
      $this->front = $this->front->next;

      if($this->front == null){
        $this->rear = null;
      }

      echo $front->data." is deenqued. ";

      return $front;
    }
    else{
      echo "Queue is empty";
      return ;
    }
  }

  public function rear(){
    return $this->rear;
  }

  public function front(){
    return $this->front;
  }

}

$queue = new Queue();
$queue->enqueue('20');
$queue->enqueue('30');
$queue->enqueue('40');
$queue->dequeue();
$queue->enqueue('50');
$queue->dequeue();

File: cracking_the_coding_interview/stack_queue/golang/queue.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type Queue struct {
	front *Node
	rear  *Node
}

func (q *Queue) EnQueue(data int) {
	n := &Node{data: data}

	if q.front == nil {
		q.front = n
	}

	if q.rear == nil {
		q.rear = n
		return
	}

	q.rear.next = n
	q.rear = n
}

func (q *Queue) IsEmpty() bool {
	return q.front == nil
}

func (q *Queue) DeQueue() int {
	if q.front == nil {
		return -1
	}

	data := q.front.data
	q.front = q.front.next

	if q.front == nil {
		q.rear = nil
	}

	return data
}

func main() {

	queue := Queue{}
	queue.EnQueue(10)
	queue.EnQueue(10)
	queue.EnQueue(13)
	queue.EnQueue(10)

	fmt.Println("Dequeue:", queue.DeQueue())
	fmt.Println("Dequeue:", queue.DeQueue())
	fmt.Println("Dequeue:", queue.DeQueue())

}
